File: util/pack/WaveAnim.java
package common.util.pack;

import common.system.fake.FakeImage;
import common.util.anim.AnimI;
import common.util.anim.EAnimD;
import common.util.anim.MaAnim;
import common.util.anim.MaModel;

public class WaveAnim extends AnimI<WaveAnim, WaveAnim.WaveType> {

	public enum WaveType implements AnimI.AnimType<WaveAnim, WaveType> {
		DEF
	}

	private final Background bg;
	private final MaModel mamodel;
	private final MaAnim maanim;

	private FakeImage[] parts;

	public WaveAnim(Background BG, MaModel model, MaAnim anim) {
		bg = BG;
		mamodel = model;
		maanim = anim;
	}

	@Override
	public void check() {
		if (parts == null)
			load();
	}

	@Override
	public EAnimD<WaveType> getEAnim(WaveType t) {
		return new EAnimD<>(this, mamodel, maanim, t);
	}

	@Override
	public void load() {
		bg.check();
		parts = bg.parts;
	}

	@Override
	public String[] names() {
		return translate(WaveType.DEF);
	}

	@Override
	public FakeImage parts(int i) {
		check();
		return parts[i];
	}

	@Override
	public WaveType[] types() {
		return WaveType.values();
	}

}

File: util/pack/WaveAnim.kt
package common.util.pack

import common.system.fake.FakeImage
import common.util.anim.AnimI
import common.util.anim.EAnimD
import common.util.anim.MaAnim
import common.util.anim.MaModel
import common.util.pack.WaveAnim.WaveType

class WaveAnim(private val bg: Background, model: MaModel, anim: MaAnim) : AnimI<WaveAnim?, WaveType?>() {
    enum class WaveType : AnimType<WaveAnim?, WaveType?> {
        DEF
    }

    private val mamodel: MaModel
    private val maanim: MaAnim
    private var parts: Array<FakeImage>?
    override fun check() {
        if (parts == null) load()
    }

    override fun getEAnim(t: WaveType): EAnimD<WaveType>? {
        return EAnimD<WaveType>(this, mamodel, maanim)
    }

    override fun load() {
        bg.check()
        parts = bg.parts
    }

    override fun names(): Array<String?> {
        return AnimI.Companion.translate(WaveType.DEF)
    }

    override fun parts(i: Int): FakeImage? {
        check()
        return parts!![i]
    }

    override fun types(): Array<WaveType> {
        return common.util.pack.WaveAnim.WaveType.values()
    }

    init {
        mamodel = model
        maanim = anim
    }
}
